lib/sitemap/generator.rb

```diff
@@ -0,0 +1,150 @@
+module Sitemap
+  class Generator
+    def initialize(search_config, uploader, timestamp)
+      @search_config    = search_config
+      @search_client    = Services.elasticsearch(hosts: search_config.base_uri, timeout: 10)
+      @uploader         = uploader
+      @timestamp        = timestamp
+    end
+
+    def run
+      create_sitemap_index(create_sitemaps(get_all_documents))
+    end
+
+    def create_sitemaps(enumerator)
+      enumerator.with_index.map do |documents, index|
+        batch_number = index + 1
+        documents.unshift(homepage) if batch_number == 1
+        create_sitemap(documents, batch_number)
+      end
+    end
+
+    def create_sitemap(documents, batch_number)
+      file_name = "sitemap_#{batch_number}.xml"
+      @uploader.upload(
+        file_content: generate_sitemap_xml(documents),
+        file_name:    file_name,
+      )
+      file_name
+    end
+
+    def create_sitemap_index(sitemaps)
+      @uploader.upload(
+        file_content: generate_sitemap_index_xml(sitemaps),
+        file_name:    "sitemap.xml",
+      )
+    end
+
+    # Generate a sitemap which matches the format specified in https://www.sitemaps.org/protocol.html
+    def generate_sitemap_xml(documents)
+      builder = Nokogiri::XML::Builder.new(encoding: "UTF-8") do |xml|
+        xml.urlset(xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9") do
+          documents.each do |document|
+            xml.url {
+              xml.loc document.url
+              xml.lastmod document.last_updated if document.last_updated
+              xml.priority document.priority
+            }
+          end
+        end
+      end
+      builder.to_xml
+    end
+
+    def generate_sitemap_index_xml(sitemap_filenames)
+      builder = Nokogiri::XML::Builder.new(encoding: "UTF-8") do |xml|
+        xml.sitemapindex(xmlns: "http://www.sitemaps.org/schemas/sitemap/0.9") do
+          sitemap_filenames.each do |sitemap_filename|
+            xml.sitemap {
+              xml.loc "#{base_url}/#{SUB_DIRECTORY}/#{sitemap_filename}"
+              xml.lastmod @timestamp.strftime("%FT%T%:z")
+            }
+          end
+        end
+      end
+      builder.to_xml
+    end
+
+  private
+
+    def get_all_documents
+      page                  = initial_scroll
+      scroll_id             = page["_scroll_id"]
+      documents             = page["hits"]["hits"]
+      total_documents_count = page["hits"]["total"]
+      total_page_count      = total_documents_count.fdiv(SCROLL_BATCH_SIZE).ceil
+      chunk                 = []
+
+      Enumerator::Lazy.new(0...total_page_count) do |yielder|
+        documents.map do |document|
+          chunk << SitemapPresenter.new(document["_source"], property_boost_calculator)
+
+          if chunk.size == SITEMAP_LIMIT
+            yielder << chunk
+            chunk = []
+          end
+        end
+
+        documents = scroll(scroll_id)
+
+        if documents.empty?
+          yielder << chunk
+        end
```

Comment:
If we get multiple empty `documents` lists at the end of the scroll (I don't know if that can happen), then this will yield the same chunk multiple times.  So we should do something like

```ruby
if documents.empty?
  yielder << chunk unless chunk.empty?
  chunk = []
end
```